accepted/00039-content-lifecycle-management.md

```diff
@@ -171,18 +175,57 @@ similar to the schedule definition of a repository sync.
+
+Note: for Software Channels, the restore functionality does not roll back installed packages. It only affects contents of the channels.
+
+The most important entity is the *Snapshot*. It has a version and it contains all the information needed for re-creating a *Build* in particular state in the future:
+* software channels reference (also the pointer which channel is the base (or "leader) in the *Build*
+* packages per channel reference
+* errata per channel
+* others (after we implement support for them): activation key, configuration channels, ...
+
+### Create Content Snapshot (backup)
+*Snapshot* is created after a successful *Project* *build*. It is populated with the references to channels and their packages/errata based on the *build*. Note that the software channels references in the snapshot point to the **source** channels of a project, not to the built ones.
+
+#### Example
+
+Project (`prj1` label) source channels:
+`sle-15-sp1-pool(packages: kernel, systemd, cron), sle-15-sp1-update(packages: kernel, systemd)`
+
+Channels in the 1st environment (`dev` label):
+`prj1-dev-sle-15-sp1-pool(packages: kernel, systemd), prj1-dev-sle-15-sp1-update(packages: kernel, systemd)` (the `cron` packages was filtered out with a filter)
+
+Create *Snapshot* of the `dev` *Environment*, it contains:
+- references to **source** channels: `sle-15-sp1-pool` and `sle-15-sp1-update`,
+- references to packages contained in the 1st *Environment*: `kernel` and `systemd`.
+
+
+### Restore Content Snapshot
+Restoring a *Snapshot* to a *Environment*.
+
+The data from the *Snapshot* is evaluated and the *Content* is re-created from it (e.g. the software channel tree is re-created in given environment and populated by packages and errata from the *Snapshot*).
 
-As the current design does not allow a rollback of an *Environment* to previous versions of the *Content Build* the only chance to restore an *Environment* is to use the one from one stage above (restore).
+Technical note: For this, the adapted code for promoting *Environment*s could be used.
 
-Example scenario:
-* 3 *Environment*s: `DEV`, `TEST`, `PRODUCTION`
-* Promoting *Content Build* from `DEV` to `TEST`
-* Testing the patches on systems in `TEST` fails, now we want to bring back the previous state:
 
-The solution can be:
- * 'restore' from `PRODUCTION` to `TEST`
+### Version management
+Keeping all versions of *Snapshot*s throughout the *Project* existence would be impractical for 2 reasons:
+* DB pollution (growing table for *Snapshot*)
+* integrity issues (it is likely that some entities referenced from very old *Snapshot*s will be deleted).
 
+The proposal is to keep only selected subset of the *Snapshot*s. There are 2 kinds of snapshots: impotant ones and other (unimportant).
+
+By default, history of 3 important *Snapshot*s and 1 unimportant one is kept (these numbers can be overriden in Uyuni global settings). *Snapshot* of current state of *Environment*s are kept by default.
```

Comment:
Keeping "3 important" is easy, as we take the overall number and remove the "oldest" or lowest version if we have more than 3.

Keeping "1 unimportant" is not so clear and we should specify it a bit more how the algorithm is.
The reason is, that we will have more than 1 unimportant snapshot as we need to keep all which are materialized in Environments.

I think we should say we keep snapshots in addition to the materialized snapshots in the environments.

This would mean for the algorithm:

- take all unimportant snapshots
- remove the versions which are build in one of the environments
- from the leftovers remove the oldest/lowest versions until we reach the configured number (e.g. 1)

(Same for the important snapshots.)

```suggestion
By default, history of 3 important *Snapshot*s and 1 unimportant one is kept in addition to the materialized versions (these numbers can be overriden in Uyuni global settings). *Snapshot* of current state of *Environment*s are kept by default.
```